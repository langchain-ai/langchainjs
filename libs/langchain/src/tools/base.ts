export {
  type ToolParams,
  ToolInputParsingException,
  StructuredTool,
  Tool,
} from "@langchain/core/tools";
