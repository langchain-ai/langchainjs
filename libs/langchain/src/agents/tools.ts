import type {
  StructuredToolInterface,
  DynamicTool,
} from "@langchain/core/tools";
import type { RunnableToolLike } from "@langchain/core/runnables";

export type ServerTool = Record<string, unknown>;
export type ClientTool =
  | StructuredToolInterface
  | DynamicTool
  | RunnableToolLike;
