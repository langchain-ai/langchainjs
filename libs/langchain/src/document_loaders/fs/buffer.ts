import type { readFile as ReadFileT } from "node:fs/promises";
import { Document } from "@langchain/core/documents";
import { getEnv } from "@langchain/core/utils/env";
import { BaseDocumentLoader } from "@langchain/core/document_loaders/base";

/**
 * Abstract class that extends the `BaseDocumentLoader` class. It
 * represents a document loader that loads documents from a buffer. The
 * `load()` method is implemented to read the buffer contents and metadata
 * based on the type of `filePathOrBlob`, and then calls the `parse()`
 * method to parse the buffer and return the documents.
 */
export abstract class BufferLoader extends BaseDocumentLoader {
  constructor(public filePathOrBlob: string | Blob) {
    super();
  }

  /**
   * Abstract method that needs to be implemented by subclasses. It is used
   * to parse the buffer and return the documents.
   * @param raw The buffer to be parsed.
   * @param metadata Metadata of the document.
   * @returns Promise that resolves with an array of `Document` objects.
   */
  protected abstract parse(
    raw: Buffer,
    metadata: Document["metadata"]
  ): Promise<Document[]>;

  /**
   * Method that reads the buffer contents and metadata based on the type of
   * `filePathOrBlob`, and then calls the `parse()` method to parse the
   * buffer and return the documents.
   * @returns Promise that resolves with an array of `Document` objects.
   */
  public async load(): Promise<Document[]> {
    let buffer: Buffer;
    let metadata: Record<string, string>;
    if (typeof this.filePathOrBlob === "string") {
      const { readFile } = await BufferLoader.imports();
      buffer = await readFile(this.filePathOrBlob);
      metadata = { source: this.filePathOrBlob };
    } else {
      buffer = await this.filePathOrBlob
        .arrayBuffer()
        .then((ab) => Buffer.from(ab));
      metadata = { source: "blob", blobType: this.filePathOrBlob.type };
    }
    return this.parse(buffer, metadata);
  }

  /**
   * Static method that imports the `readFile` function from the
   * `fs/promises` module in Node.js. It is used to dynamically import the
   * function when needed. If the import fails, it throws an error
   * indicating that the `fs/promises` module is not available in the
   * current environment.
   * @returns Promise that resolves with an object containing the `readFile` function.
   */
  static async imports(): Promise<{
    readFile: typeof ReadFileT;
  }> {
    try {
      const { readFile } = await import("node:fs/promises");
      return { readFile };
    } catch (e) {
      console.error(e);
      throw new Error(
        `Failed to load fs/promises. TextLoader available only on environment 'node'. It appears you are running environment '${getEnv()}'. See https://<link to docs> for alternatives.`
      );
    }
  }
}
