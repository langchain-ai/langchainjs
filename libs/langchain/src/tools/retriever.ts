import type { BaseRetrieverInterface } from "@langchain/core/retrievers";
import { z } from "zod";
import { CallbackManagerForToolRun } from "@langchain/core/callbacks/manager";
import {
  DynamicStructuredTool,
  type DynamicStructuredToolInput,
} from "@langchain/core/tools";
import { formatDocumentsAsString } from "../util/document.js";

export function createRetrieverTool(
  retriever: BaseRetrieverInterface,
  input: Omit<DynamicStructuredToolInput, "func" | "schema">
) {
  const func = async (
    { query }: { query: string },
    runManager?: CallbackManagerForToolRun
  ) => {
    const docs = await retriever.getRelevantDocuments(
      query,
      runManager?.getChild("retriever")
    );
    return formatDocumentsAsString(docs);
  };
  const schema = z.object({
    query: z.string().describe("query to look up in retriever"),
  });
  return new DynamicStructuredTool({ ...input, func, schema });
}
