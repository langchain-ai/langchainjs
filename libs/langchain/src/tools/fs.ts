import { z } from "zod/v3";
import { StructuredTool, ToolParams } from "@langchain/core/tools";
import { InferInteropZodOutput } from "@langchain/core/utils/types";
import { BaseFileStore } from "../stores/file/base.js";

/**
 * Interface for parameters required by the ReadFileTool class.
 */
interface ReadFileParams extends ToolParams {
  store: BaseFileStore;
}

const readSchema = z.object({
  file_path: z.string().describe("name of file"),
});
type ReadToolSchema = typeof readSchema;

/**
 * Class for reading files from the disk. Extends the StructuredTool
 * class.
 */
export class ReadFileTool extends StructuredTool {
  static lc_name() {
    return "ReadFileTool";
  }

  schema = readSchema;

  name = "read_file";

  description = "Read file from disk";

  store: BaseFileStore;

  constructor({ store }: ReadFileParams) {
    super(...arguments);

    this.store = store;
  }

  async _call({ file_path }: InferInteropZodOutput<ReadToolSchema>) {
    return await this.store.readFile(file_path);
  }
}

/**
 * Interface for parameters required by the WriteFileTool class.
 */
interface WriteFileParams extends ToolParams {
  store: BaseFileStore;
}

const writeSchema = z.object({
  file_path: z.string().describe("name of file"),
  text: z.string().describe("text to write to file"),
});
type WriteToolSchema = typeof writeSchema;

/**
 * Class for writing data to files on the disk. Extends the StructuredTool
 * class.
 */
export class WriteFileTool extends StructuredTool {
  static lc_name() {
    return "WriteFileTool";
  }

  schema = writeSchema;

  name = "write_file";

  description = "Write file from disk";

  store: BaseFileStore;

  constructor({ store, ...rest }: WriteFileParams) {
    super(rest);

    this.store = store;
  }

  async _call({ file_path, text }: InferInteropZodOutput<WriteToolSchema>) {
    await this.store.writeFile(file_path, text);
    return "File written to successfully.";
  }
}
