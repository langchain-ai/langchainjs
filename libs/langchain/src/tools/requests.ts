import { Tool } from "@langchain/core/tools";

export interface Headers {
  [key: string]: string;
}

/**
 * Interface for HTTP request tools. Contains properties for headers and
 * maximum output length.
 */
export interface RequestTool {
  headers: Headers;
  maxOutputLength: number;
}

/**
 * Class for making GET requests. Extends the Tool class and implements
 * the RequestTool interface. The input should be a URL string, and the
 * output will be the text response of the GET request.
 */
export class RequestsGetTool extends Tool implements RequestTool {
  static lc_name() {
    return "RequestsGetTool";
  }

  name = "requests_get";

  maxOutputLength = 2000;

  constructor(
    public headers: Headers = {},
    { maxOutputLength }: { maxOutputLength?: number } = {}
  ) {
    super(...arguments);

    this.maxOutputLength = maxOutputLength ?? this.maxOutputLength;
  }

  /** @ignore */
  async _call(input: string) {
    const res = await fetch(input, {
      headers: this.headers,
    });
    const text = await res.text();
    return text.slice(0, this.maxOutputLength);
  }

  description = `A portal to the internet. Use this when you need to get specific content from a website.
  Input should be a url string (i.e. "https://www.google.com"). The output will be the text response of the GET request.`;
}

/**
 * Class for making POST requests. Extends the Tool class and implements
 * the RequestTool interface. The input should be a JSON string with two
 * keys: 'url' and 'data'. The output will be the text response of the
 * POST request.
 */
export class RequestsPostTool extends Tool implements RequestTool {
  static lc_name() {
    return "RequestsPostTool";
  }

  name = "requests_post";

  maxOutputLength = Infinity;

  constructor(
    public headers: Headers = {},
    { maxOutputLength }: { maxOutputLength?: number } = {}
  ) {
    super(...arguments);

    this.maxOutputLength = maxOutputLength ?? this.maxOutputLength;
  }

  /** @ignore */
  async _call(input: string) {
    try {
      const { url, data } = JSON.parse(input);
      const res = await fetch(url, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(data),
      });
      const text = await res.text();
      return text.slice(0, this.maxOutputLength);
    } catch (error) {
      return `${error}`;
    }
  }

  description = `Use this when you want to POST to a website.
  Input should be a json string with two keys: "url" and "data".
  The value of "url" should be a string, and the value of "data" should be a dictionary of
  key-value pairs you want to POST to the url as a JSON body.
  Be careful to always use double quotes for strings in the json string
  The output will be the text response of the POST request.`;
}
