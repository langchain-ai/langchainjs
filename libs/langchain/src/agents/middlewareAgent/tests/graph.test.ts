import fs from "node:fs";
import { describe, it, expect, afterAll } from "vitest";
import { tool } from "@langchain/core/tools";
import { createAgent, createMiddleware } from "../index.js";
import type { JumpToTarget } from "../types.js";

// Strategic test cases covering meaningful middleware interaction patterns
// rather than exhaustive combinations which would result in 13,122 test cases

type Case = {
  aBefore?: JumpToTarget[];
  aAfter?: JumpToTarget[];
  bBefore?: JumpToTarget[];
  bAfter?: JumpToTarget[];
  hasTool: boolean;
};

// Strategic matrix covering meaningful interaction patterns
const strategicCases: Omit<Case, "hasTool">[] = [
  // Basic cases - no jumps
  {
    aBefore: undefined,
    aAfter: undefined,
    bBefore: undefined,
    bAfter: undefined,
  },
  { aBefore: [], aAfter: [], bBefore: [], bAfter: [] },

  // Single middleware patterns
  {
    aBefore: ["tools"],
    aAfter: undefined,
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["tools"],
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: ["model"],
    aAfter: undefined,
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["model"],
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: ["end"],
    aAfter: undefined,
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["end"],
    bBefore: undefined,
    bAfter: undefined,
  },

  // Interaction patterns - both middleware active
  {
    aBefore: ["tools"],
    aAfter: undefined,
    bBefore: ["model"],
    bAfter: undefined,
  },
  {
    aBefore: ["model"],
    aAfter: undefined,
    bBefore: ["tools"],
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["tools"],
    bBefore: undefined,
    bAfter: ["model"],
  },
  {
    aBefore: undefined,
    aAfter: ["model"],
    bBefore: undefined,
    bAfter: ["tools"],
  },

  // Sequential dependency patterns
  {
    aBefore: ["tools"],
    aAfter: ["model"],
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: undefined,
    bBefore: ["tools"],
    bAfter: ["model"],
  },
  {
    aBefore: ["model"],
    aAfter: ["tools"],
    bBefore: undefined,
    bAfter: undefined,
  },

  // Termination patterns
  {
    aBefore: ["end"],
    aAfter: undefined,
    bBefore: ["tools"],
    bAfter: undefined,
  },
  {
    aBefore: ["tools"],
    aAfter: ["end"],
    bBefore: ["model"],
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["end"],
    bBefore: undefined,
    bAfter: ["tools"],
  },

  // Multiple option patterns
  {
    aBefore: ["tools", "model"],
    aAfter: undefined,
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: undefined,
    aAfter: ["tools", "model"],
    bBefore: undefined,
    bAfter: undefined,
  },
  {
    aBefore: ["tools", "end"],
    aAfter: undefined,
    bBefore: ["model"],
    bAfter: undefined,
  },
  {
    aBefore: ["model", "end"],
    aAfter: undefined,
    bBefore: ["tools"],
    bAfter: undefined,
  },

  // Complex interaction patterns
  {
    aBefore: ["tools", "model"],
    aAfter: ["end"],
    bBefore: ["tools"],
    bAfter: ["model"],
  },
  {
    aBefore: ["tools"],
    aAfter: ["model", "end"],
    bBefore: ["model"],
    bAfter: ["tools"],
  },
  {
    aBefore: ["tools", "model", "end"],
    aAfter: undefined,
    bBefore: undefined,
    bAfter: ["tools", "model", "end"],
  },

  // Edge cases - conflicting or unusual patterns
  { aBefore: ["end"], aAfter: ["tools"], bBefore: ["end"], bAfter: ["model"] },
  {
    aBefore: ["tools", "model", "end"],
    aAfter: ["tools", "model", "end"],
    bBefore: ["tools", "model", "end"],
    bAfter: ["tools", "model", "end"],
  },
];

const hasToolOptions = [false, true];
const matrix: Case[] = hasToolOptions.flatMap((hasTool) =>
  strategicCases.map((testCase) => ({
    ...testCase,
    hasTool,
  }))
);

const collected: {
  aBeforeLabel: string;
  aAfterLabel: string;
  bBeforeLabel: string;
  bAfterLabel: string;
  toolsLabel: string;
  mermaid: string;
}[] = [];

const someTool = tool(async () => "Hello, world!", {
  name: "someTool",
});

describe.each(matrix)(
  "graph (A before: $aBefore, A after: $aAfter | B before: $bBefore, B after: $bAfter | tools: $hasTool)",
  ({ aBefore, aAfter, bBefore, bAfter, hasTool }) => {
    it("should create correct graph structure", async () => {
      const middleware1 = createMiddleware({
        name: "MiddlewareA",
        beforeModelJumpTo: aBefore,
        afterModelJumpTo: aAfter,
        beforeModel: () => {},
        afterModel: () => {},
      });

      const middleware2 = createMiddleware({
        name: "MiddlewareB",
        beforeModelJumpTo: bBefore,
        afterModelJumpTo: bAfter,
        beforeModel: () => {},
        afterModel: () => {},
      });

      const agent = createAgent({
        model: "openai:gpt-4o-mini",
        tools: hasTool ? [someTool] : [],
        middleware: [middleware1, middleware2] as const,
      });

      expect(agent.graph).toMatchSnapshot();
      const fmt = (v?: JumpToTarget[]) =>
        v === undefined
          ? "undefined"
          : v.length === 0
          ? "empty"
          : `<code>${v.join("</code>, <code>")}</code>`;
      const aBeforeLabel = fmt(aBefore);
      const aAfterLabel = fmt(aAfter);
      const bBeforeLabel = fmt(bBefore);
      const bAfterLabel = fmt(bAfter);
      const toolsLabel = hasTool ? "yes" : "no";
      const mermaid = await agent.drawMermaid({ withStyles: false });
      collected.push({
        aBeforeLabel,
        aAfterLabel,
        bBeforeLabel,
        bAfterLabel,
        toolsLabel,
        mermaid,
      });
    });
  }
);

afterAll(() => {
  const content = `# Graph Matrix (Mermaid)\n\n
This test creates a matrix of all possible combinations of before and after model jump targets for two middleware instances.
The basic test setup is as follows:

\`\`\`ts
const middleware1 = createMiddleware({
    name: "MiddlewareA",
    beforeModelJumpTo: aBefore,
    afterModelJumpTo: aAfter,
    beforeModel: () => {},
    afterModel: () => {},
});

const middleware2 = createMiddleware({
    name: "MiddlewareB",
    beforeModelJumpTo: bBefore,
    afterModelJumpTo: bAfter,
    beforeModel: () => {},
    afterModel: () => {},
});

const agent = createAgent({
    model: "openai:gpt-4o-mini",
    tools: [someTool],
    middleware: [middleware1, middleware2] as const,
});
\`\`\`

${collected
  .map(
    ({
      aBeforeLabel,
      aAfterLabel,
      bBeforeLabel,
      bAfterLabel,
      toolsLabel,
      mermaid,
    }) =>
      `<details>
<summary>

MiddlewareA before: ${aBeforeLabel}<br/>
MiddlewareA after: ${aAfterLabel}<br/>
MiddlewareB before: ${bBeforeLabel}<br/>
MiddlewareB after: ${bAfterLabel}<br/>
Tools: <code>${toolsLabel}</code>

</summary>

\`\`\`mermaid
${mermaid}
\`\`\`

</details>`
  )
  .join("\n\n")}`;
  fs.writeFileSync("graph-matrix.mermaid.md", content);
});
