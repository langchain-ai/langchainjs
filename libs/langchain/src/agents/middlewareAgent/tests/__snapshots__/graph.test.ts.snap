// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graph (A before: [ 'end' ], A after: [ 'tools' ] | B before: [ 'end' ], B after: [ 'model' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'end' ], A after: [ 'tools' ] | B before: [ 'end' ], B after: [ 'model' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'end' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'end' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'end' ], A after: undefined | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'end' ], A after: undefined | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: [ 'tools' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: [ 'tools' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: undefined | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model' ], A after: undefined | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model', 'end' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'model', 'end' ], A after: undefined | B before: [ 'tools' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'end' ] | B before: [ 'model' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'end' ] | B before: [ 'model' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'model' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'model' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'model', 'end' ] | B before: [ 'model' ], B after: [ 'tools' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: [ 'model', 'end' ] | B before: [ 'model' ], B after: [ 'tools' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: undefined | B before: [ 'model' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: undefined | B before: [ 'model' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: undefined | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools' ], A after: undefined | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'end' ], A after: undefined | B before: [ 'model' ], B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'end' ], A after: undefined | B before: [ 'model' ], B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model' ], A after: [ 'end' ] | B before: [ 'tools' ], B after: [ 'model' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model' ], A after: [ 'end' ] | B before: [ 'tools' ], B after: [ 'model' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model' ], A after: undefined | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model' ], A after: undefined | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model', 'end' ], A after: [ 'tools', 'model', 'end' ] | B before: [ 'tools', 'model', 'end' ], B after: [ 'tools', 'model', 'end' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model', 'end' ], A after: [ 'tools', 'model', 'end' ] | B before: [ 'tools', 'model', 'end' ], B after: [ 'tools', 'model', 'end' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model', 'end' ], A after: undefined | B before: undefined, B after: [ 'tools', 'model', 'end' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [ 'tools', 'model', 'end' ], A after: undefined | B before: undefined, B after: [ 'tools', 'model', 'end' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [], A after: [] | B before: [], B after: [] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: [], A after: [] | B before: [], B after: [] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'end' ] | B before: undefined, B after: [ 'tools' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'end' ] | B before: undefined, B after: [ 'tools' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'end' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'end' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'model' ] | B before: undefined, B after: [ 'tools' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'model' ] | B before: undefined, B after: [ 'tools' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'model' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'model' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools' ] | B before: undefined, B after: [ 'model' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools' ] | B before: undefined, B after: [ 'model' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools', 'model' ] | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: [ 'tools', 'model' ] | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: undefined | B before: [ 'tools' ], B after: [ 'model' ] | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: undefined | B before: [ 'tools' ], B after: [ 'model' ] | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: undefined | B before: undefined, B after: undefined | tools: false) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;

exports[`graph (A before: undefined, A after: undefined | B before: undefined, B after: undefined | tools: true) > should create correct graph structure 1`] = `
{
  "id": [
    "langgraph",
    "pregel",
    "CompiledStateGraph",
  ],
  "lc": 1,
  "type": "not_implemented",
}
`;
